pkg/external-provider/external_series_registry.go

```diff
@@ -0,0 +1,109 @@
+package provider
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/provider"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/labels"
+
+	prom "github.com/directxman12/k8s-prometheus-adapter/pkg/client"
+)
+
+// ExternalSeriesRegistry acts as the top-level converter for transforming Kubernetes requests
+// for external metrics into Prometheus queries.
+type ExternalSeriesRegistry interface {
+	// ListAllMetrics lists all metrics known to this registry
+	ListAllMetrics() []provider.ExternalMetricInfo
+	QueryForMetric(namespace string, metricName string, metricSelector labels.Selector) (prom.Selector, bool, error)
+}
+
+// overridableSeriesRegistry is a basic SeriesRegistry
+type externalSeriesRegistry struct {
+	mu sync.RWMutex
+
+	// metrics is the list of all known metrics, ready to return from the API
+	metrics []provider.ExternalMetricInfo
+	// rawMetrics is a lookup from a metric to SeriesConverter for the sake of generating queries
+	rawMetrics map[string]SeriesConverter
+
+	mapper apimeta.RESTMapper
+
+	metricLister MetricListerWithNotification
+}
+
+// NewExternalSeriesRegistry creates an ExternalSeriesRegistry driven by the data from the provided MetricLister.
+func NewExternalSeriesRegistry(lister MetricListerWithNotification, mapper apimeta.RESTMapper) ExternalSeriesRegistry {
+	var registry = externalSeriesRegistry{
+		mapper:       mapper,
+		metricLister: lister,
+		metrics:      make([]provider.ExternalMetricInfo, 0),
+		rawMetrics:   map[string]SeriesConverter{},
+	}
+
+	lister.AddNotificationReceiver(registry.filterAndStoreMetrics)
+
+	return &registry
+}
+
+func (r *externalSeriesRegistry) filterAndStoreMetrics(result MetricUpdateResult) {
+
+	newSeriesSlices := result.series
+	converters := result.converters
+
+	if len(newSeriesSlices) != len(converters) {
+		glog.Errorf("need one set of series per converter")
```

Comment:
This will panic a few lines below if `len(newSeriesSlices) < len(converters)`. We should return with an error from this function instead.